.golangci.yml

```diff
@@ -20,6 +20,7 @@ linters-settings:
     include-go-root: false
     packages-with-error-message:
       - github.com/rotisserie/eris: "use ctxerr.New or ctxerr.Wrap[f] instead"
+      - github.com/pkg/errors: "use ctxerr if a context.Context is available or stdlib errors.New / fmt.Errorf with the %w verb"
```

Comment:
I understand this change, but it's preventing us from having stack traces when there's no context.

The ideal case would be to generate the enriched error once, with or without ctx. Everywhere else in the stack, we just return the err, no need to re-wrap or anything.

So you type the least possible, but get the most data.

Reply:
Having to type `fmt.Errorf("some written context: %w", err)` is more than just `ctxerr.Wrap(err)` or just `err`.

Reply:
> I understand this change, but it's preventing us from having stack traces when there's no context.

The idea is that some packages are "helper" packages or general-purpose ones, they will be called and used by handlers somewhere (or CLI or some other entrypoint, but I think in our case CLI commands and HTTP handlers covers the overwhelming majority of cases). That's where you will generate the stack trace when wrapping that error. It's the same as if you call e.g. `json.Marshal`, you will not generate the stack trace inside that function, but you will generate it at the call site.

> The ideal case would be to generate the enriched error once, with or without ctx. Everywhere else in the stack, we just return the err, no need to re-wrap or anything.

I disagree with that, I think there's tremendous value in annotating an error message even if the stack trace exists, it's much faster and easier to understand what's going on if there's a clear error message than if you have to investigate the call stack which requires first checking out the exact code version used and possibly locating a bunch of files and calls.

Reply:
> The idea is that some packages are "helper" packages or general-purpose ones, they will be called and used by handlers somewhere (or CLI or some other entrypoint, but I think in our case CLI commands and HTTP handlers covers the overwhelming majority of cases). That's where you will generate the stack trace when wrapping that error. It's the same as if you call e.g. `json.Marshal`, you will not generate the stack trace inside that function, but you will generate it at the call site.

Right, I guess I hadn't put any of our code in the same bag as `json.Marshal`.

> I disagree with that, I think there's tremendous value in annotating an error message even if the stack trace exists, it's much faster and easier to understand what's going on if there's a clear error message than if you have to investigate the call stack which requires first checking out the exact code version used and possibly locating a bunch of files and calls.

That's fair. My comment comes from having to use only the annotations to figure out what's going on, and it was pretty annoying/tricky, mostly because the annotations/errors are `updating foo: inserting blah into the db: Error XXXX: row is lock`. So it's not very useful, you have to grab some of that string and search it around the code to see what call failed in that way or what query exactly was the one of many in the tx to fail in that way.

I suppose there's a world where an annotation gives you all the information you need, or really adds value even if it's not all the information. I just don't think that's the world we have at fleet right now, so my inclination was to replace the annotations altogether for something that pin points the error faster.

> which requires first checking out the exact code version used and possibly locating a bunch of files and calls

This is no different than what we have currently, unless you assume the annotation is really useful and doesn't need you to search it around the code, which I haven't seen at all so far. I've seen it used more like a "you've passed through here" tag.

Reply:
> Right, I guess I hadn't put any of our code in the same bag as json.Marshal.

That's just an example, capturing stack traces must start somewhere and I thought having it start where you have the context (which is typically the call stack that is interesting in debugging) is the right place for this. You can't really just forget about wrapping and return a naked `err`, as that means checking the implementation of everything you call to make sure it does indeed capture what you want.


> I suppose there's a world where an annotation gives you all the information you need, or really adds value even if it's not all the information. I just don't think that's the world we have at fleet right now

I don't know, to me something like `delete existing scheduled queries: Error XXXX: row is locked` is way more useful at a glance than say `server/datastore/mysql/packs.go:58:10`. It's a bit more than "you've passed through here", it's "that specific action caused that error".

I guess I'll put that aside for now, I don't want to go back and forth changing a bunch of error handling code without making sure we have a clear plan/target - I mentioned what I was going for in this PR's comment here https://github.com/fleetdm/fleet/pull/2786#issuecomment-963255877, let's first agree on an approach before I change this PR as it's significant changes.

Reply:
> let's first agree on an approach before I change this PR as it's significant changes.

YES! Definitely, not meaning to make you go back and forth, just bouncing off ideas.

Reply:
> I don't know, to me something like `delete existing scheduled queries: Error XXXX: row is locked` is way more useful at a glance than say `server/datastore/mysql/packs.go:58:10`. It's a bit more than "you've passed through here", it's "that specific action caused that error".

Sure, annotating + error message + stack trace is better than just either one of them.

As for when we start the trace, that's fine. If we find a limitation we can move the starting point somewhere else.